### PR TITLE
Analytics: Add missing os-type cases, and make an unset os-type an error.

### PR DESCRIPTION
**Pull request by waddlesplash**
The rationale for the #error is as follows (because there was some controversy on #dolphin-dev):

 * I didn't know about this #if block when I did the initial Haiku port (#4951), and nobody reminded me
 * It should not break the build anywhere it presently works, and for future platforms it's helpful because otherwise you have to spend some time finding it (like I did) or miss it entirely (like I did, at first)

Although I don't even know where else there is to port Dolphin to. IllumOS? (lol)

### CODE REVIEW

**Comment by Helios747** [comment]
Don't agree with changing that to #error. IMO #error should only be used when it would cause a problem at runtime. Having an "unknown" in analytics is whatever.

**Comment by waddlesplash** [comment]
Should this get documented somewhere, then? Dolphin does not presently have a "porting" guide, and I had to do some hunting to locate it. #error seems to me the easiest way to handle this from just about any perspective, and it's not much hassle, really.

**Comment by sepalani** [comment]
If OpenBSD happened to work before, that ```#error``` usage might prevent other distros (could be active forks) to build Dolphin even if there are capable of running it. Moreover, that part of the code is related to Analytics that isn't used by everyone and shouldn't prevent Dolphin to compile.

Last but not least, ```#error``` is mostly used in [Core/MachineContext.h](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/MachineContext.h) and I don't think it's a good idea to spread its usage where it doesn't harm. Affected users might not even care or be able to fix it if they aren't developers.

**Comment by waddlesplash** [comment]
Added NetBSD.

**Comment by BhaaLseN** [comment]
I'd leave the type "unknown" and only raise a message (or perhaps even warning? do we have warnings-as-error on?). It's certainly not an error case, although we'd want to extend that list whenever necessary (but as you noticed, there is no real "porting" guide anywhere that notes this).
Or, can we have CMake write the detected OS into a variable to use that one instead for the unknown case?

**Comment by degasus** [comment]
It should at least raise a warning, I'm very fine with an error, too.

**Comment by sepalani** [comment]
I'm against an error for the reasons I mentioned above and for a usage of the error preprocessor directive that I don't consider to be valid, IMO. Haiku was working without errors (even if there was missing Analytics) and the same goes for every missed/missing case mentioned above.

There were missing cases (I don't blame anyone for them as it's hard to keep track of all of them without missing one or newcomers like Haiku), where these cases were known to be working/compiling. I'm fine with a warning if it doesn't prevent the project to compile as we (maybe just I?) don't know if the maintainer of Dolphin for these is familiar to Dolphin codebase or not. And therefore might not care or might not know how to fix the error.

**Comment by BhaaLseN** [comment]
At the very least, if you insist on keeping the `#error` (which I'd very much prefer as `#warning` with some sort of fallback; either "unknown" or autodetected/provided by CMake or whatever) you should (must?) add some comments/hints on where/when `os-type` matters, what it does and why someone should care about adding theirs as new `#elseif` directive. Right now it isn't really helpful just saying "hey, os-type is not defined".

**Comment by shuffle2** [comment]
we don't actually support these OS afaik. Also, warnings are errors (or should be - if not then your toolchain is wrong :') ) so warning vs. error is just semantics.

**Comment by sepalani** [comment]
According to the Dolphin's System Requirement:
> Unix-like systems other than Linux are not officially supported but might work.

Haiku doesn't even fit this criteria as BeOS based but seems supported. Pretty sure some among others can have the same fate.

**Comment by waddlesplash** [comment]
> Haiku doesn't even fit this criteria as BeOS based but seems supported.

BeOS implemented most of POSIX, as does Haiku. (Actually we implement more of POSIX than macOS does, I think...) And we ship with Bash as the default shell, and GNU coreutils, etc. So, uh, we're UNIX-like.

**Comment by sepalani** [comment]
@waddlesplash 
POSIX is a programming interface, Unix is an Operating System/kernel such as MINIX, Unix, Linux, etc. So no, I don't think BeOS/Haiku are Unix-like but feel free to correct me if I'm wrong. Especially, if they were in fact based on it. Moreover, if that bullet point was implying Unix's POSIX compliance the bullet point for Linux would have not be needed. Last but not least, there is also a bullet point for macOS. That's my interpretation of "Unix-like", will Windows be Unix-like if it becomes POSIX compliant: I don't think so.

I'm not qualified to state that Haiku is Unix-like or not but based on what I read it doesn't seem to be one:
https://en.wikipedia.org/wiki/List_of_operating_systems#Non-Unix-like

TL;DR
My point was: Do we need to build this wall of discrimination towards unknowns supporting Dolphin by preventing them to compile?

**Comment by waddlesplash** [comment]
> Unix is an Operating System/kernel such as MINIX, Unix, Linux, etc.

Correction: UNIX was a single operating system developed at Bell Labs. All of those other OSes were "[UNIX-like](https://en.wikipedia.org/wiki/Unix-like)".

> Especially, if they were in fact based on it.

Only the BSDs were based in any real way on the original UNIX; Linux and all the rest are complete reimplementations, as is Haiku.

> Moreover, if that bullet point was implying Unix's POSIX compliance the bullet point for Linux would have not be needed.

The POSIX specification is essentially a formal definition of the APIs that UNIX had, and that Linux, *BSD, macOS et al implement. The bullet point is indeed loosely worded.

> Last but not least, there is also a bullet point for macOS.

Yes, macOS has some really weird quirks and lots of missing major POSIX functionality (last I checked, anyway?). So it is indeed a special case here.

**Comment by waddlesplash** [comment]
>  My point was: Do we need to build this wall of discrimination towards unknowns supporting Dolphin by preventing them to compile?

I guess at this point I can make it a pragma message.
